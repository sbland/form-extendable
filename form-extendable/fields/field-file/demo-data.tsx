import React from 'react';
import { IPopupProps } from '@form-extendable/lib';
import {
  EFileType,
  EFilterType,
  IFile,
} from '@react_db_client/constants.client-types';
import { IFieldFileProps } from './field-file';
import { IFieldFileMultipleProps } from './field-file-multiple';

export const DEMO_IMAGE_FILES_DATA: IFile[] = [
  {
    uid: 'a',
    filePath: '',
    label: 'FileA',
    name: 'bit-logo.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'b',
    filePath: 'extensions-icons',
    label: 'FileB',
    name: 'vue_grey.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'c',
    filePath: 'Community/app-components',
    label: 'FileC',
    name: 'card.jpg',
    fileType: EFileType.IMAGE,
  },
];

export const DEMO_FILES_DATA: IFile[] = [
  {
    uid: 'a',
    filePath: '',
    label: 'FileA with a really long name',
    name: 'bit-logo.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'b',
    filePath: 'extensions-icons',
    label: 'FileB',
    name: 'vue_grey.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'c',
    filePath: 'Community/app-components',
    label: 'FileC',
    name: 'card.jpg',
    fileType: EFileType.IMAGE,
  },
];

export const DEMO_FILES_DATA_MANY = [
  ...DEMO_FILES_DATA,
  ...Array(100)
    .fill(0)
    .map((_, i) => ({
      uid: `file_${i}`,
      filePath: '',
      label: `FileA with a really long name ${i}`,
      name: 'bit-logo.svg',
      fileType: EFileType.IMAGE,
    })),
];

export const DEMO_IMAGE_FILES_MANY = [
  ...DEMO_IMAGE_FILES_DATA,
  ...Array(100)
    .fill(0)
    .map((_, i) => ({
      uid: `file_${i}`,
      filePath: '',
      label: `FileA with a really long name ${i}`,
      name: `bit-logo_${i}.svg`,
      fileType: EFileType.IMAGE,
    })),
];

export const PopupPanel: React.FC<IPopupProps> = ({
  children,
  isOpen,
  handleClose,
}) => (isOpen ? <>{children}</> : <></>);

const onChange: IFieldFileProps['onChange'] = () => {};
const onChangeMulti: IFieldFileMultipleProps['onChange'] = () => {};

export const dummyProps: IFieldFileProps = {
  uid: 'uid',
  label: 'File Field',
  type: EFilterType.file,
  multiple: false,
  onChange,
  fileType: EFileType.IMAGE,
  value: DEMO_IMAGE_FILES_DATA[0],
  fileServerUrl: 'https://static.bit.dev',
  asyncGetFiles: (metaData) => async () => DEMO_IMAGE_FILES_DATA,
  PopupPanel,
  asyncFileUpload: () => async () => {},
};

export const dummyPropsImagesMany: IFieldFileMultipleProps = {
  ...dummyProps,
  onChange: onChangeMulti,
  type: EFilterType.fileMultiple,
  defaultValue: [],
  multiple: true,
  value: [...DEMO_IMAGE_FILES_MANY].slice(0, 50),
  asyncGetFiles: (metaData) => async () =>
    [...DEMO_IMAGE_FILES_MANY].slice(50, 60),
};

export const dummyPropsDocs: IFieldFileProps = {
  ...dummyProps,
  fileType: EFileType.DOCUMENT,
  value: DEMO_FILES_DATA[0],
  asyncGetFiles: (metaData) => async () => DEMO_FILES_DATA,
};

export const dummyPropsDocsMany: IFieldFileMultipleProps = {
  ...dummyPropsDocs,
  onChange: onChangeMulti,
  type: EFilterType.fileMultiple,
  defaultValue: [],
  multiple: true,
  value: [...DEMO_FILES_DATA_MANY].slice(0, 50),
  asyncGetFiles: (metaData) => async () =>
    [...DEMO_FILES_DATA_MANY].slice(50, 60),
};
