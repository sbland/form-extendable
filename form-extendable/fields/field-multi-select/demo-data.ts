import { IHeadingSelectMulti } from '@form-extendable/lib';
import { EFilterType } from '@react_db_client/constants.client-types';
import { TFieldMultiSelect } from './field-multi-select';

export const demoOptions = [
  {
    uid: 'a',
    label: 'A',
  },
  {
    uid: 'b',
    label: 'B',
  },
  {
    uid: 'c',
    label: 'C',
  },
];

export const demoHeading: IHeadingSelectMulti = {
  uid: 'multiSelectListShowAll',
  label: 'Multi Select Show All',
  type: EFilterType.selectMulti,
  selectType: 'showall',
  options: demoOptions,
  required: true,
};

export const props: TFieldMultiSelect = {
  type: EFilterType.selectMulti,
  label: demoHeading.label,
  uid: demoHeading.uid,
  unit: 'unit',
  onChange: () => {},
  options: demoOptions,
  value: [demoOptions[0].uid],
  required: demoHeading.required,
};
