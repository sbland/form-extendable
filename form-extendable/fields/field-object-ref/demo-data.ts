import { IObj } from '@form-extendable/lib';

export const defaultVal = '123';

export const demoResults: IObj[] = [
  {
    _id: '123',
    uid: 'foo',
    label: 'Foo',
  },
  {
    _id: '456',
    uid: 'bar',
    label: 'Bar',
  },
];
