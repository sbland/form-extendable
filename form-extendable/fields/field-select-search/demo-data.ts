export const defaultVal = { uid: 'a', label: 'A' };

export const demoOptions = [
  {
    uid: 'a',
    label: 'a',
  },
  {
    uid: 'b',
    label: 'b',
  },
  {
    uid: 'c',
    label: 'c',
  },
];
