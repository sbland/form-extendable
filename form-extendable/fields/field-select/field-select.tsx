import React, { useCallback } from 'react';
import PropTypes from 'prop-types';

import { IItem, SearchAndSelectDropdown } from '@react_db_client/components.search-and-select-dropdown';
import {
  ToggleBox,
  ToggleBoxRadioGroup,
} from '@react_db_client/components.form.form-components.toggle-box';
import { IFieldComponentProps, IHeadingSelect } from '@form-extendable/lib';

export type TFieldSelectProps = IFieldComponentProps<string> & IHeadingSelect;

export const FieldSelect = ({
  uid,
  unit,
  onChange,
  value,
  options,
  required,
  labelField = 'label',
  selectType,
  defaultValue,
  inputProps,
}: TFieldSelectProps) => {
  const selection =
    (value && options && options.find((o) => o.uid === value)) ||
    (defaultValue && options.find((o) => o.uid === defaultValue));
  const searchFunction = useCallback(async () => options, [options]);
  const valueIsInvalid = value !== null && selection === undefined;

  const selectionValidated = valueIsInvalid
    ? 'Invalid Selection'
    : selection && selection[labelField];
  const handleSelect = useCallback((data: IItem) => {
    onChange(data.uid as string);
  }, [onChange]);

  switch (selectType) {
    case 'dropdown':
      return (
        <>
          <SearchAndSelectDropdown
            searchFunction={searchFunction}
            handleSelect={handleSelect}
            initialValue={selection}
            searchFieldTargetField={labelField}
            labelField={labelField}
            className="formFieldInput"
            searchFieldPlaceholder={selectionValidated || 'search...'}
            // onChange={(e) => onChange(e.target.value)}
            required={required}
            searchDelay={0}
            allowEmptySearch
            aria-labelledby={`${uid}-label`}
            id={`${uid}-input`}
            {...inputProps}
          />
          {unit && <span>{unit}</span>}
        </>
      );
    case 'toggle':
      // TODO: Check input requirements of toggle box radio groups
      return (
        <ToggleBoxRadioGroup
          // TODO: Update toggle box
          selected={options.findIndex((opt) => opt.uid === value) as any}
          allowDeselect
          onChange={(i, v) => onChange(v as string)}
          className="flexHoriz"
        >
          {options.map((opt) => (
            <ToggleBox
              text={opt.label}
              // TODO: Update toggle box
              id={opt.uid as string}
              key={opt.uid}
              onChange={() => {}}
            />
          ))}
        </ToggleBoxRadioGroup>
      );
    default:
      return <div>INVALID SELECT TYPE {selectType} </div>;
  }
};

FieldSelect.propTypes = {
  uid: PropTypes.string.isRequired,
  unit: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  multiple: PropTypes.bool,
  required: PropTypes.bool,
  labelField: PropTypes.string,
  selectType: PropTypes.oneOf(['dropdown', 'toggle']),
  defaultVal: PropTypes.any,
};

FieldSelect.defaultProps = {
  unit: '',
  value: '',
  multiple: false,
  // defaultVal: null,
  required: false,
  labelField: 'label',
  selectType: 'dropdown',
  defaultVal: '',
} as Partial<React.ComponentProps<typeof FieldSelect>>;
