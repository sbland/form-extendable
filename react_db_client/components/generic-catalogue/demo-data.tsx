/* eslint-disable react/prop-types */
import React, { useEffect, useMemo } from 'react';
import {
  EFilterType,
  IDocument,
} from '@react_db_client/constants.client-types';
import { IHeadingButton, THeading } from '@form-extendable/lib';

export const demoHeadingsDataSimple: THeading<any>[] = [
  {
    uid: 'uid',
    label: 'UID',
    type: 'button',
    onClick: (x) => alert(x),
  } as IHeadingButton,
  {
    uid: 'name',
    label: 'Name',
    type: EFilterType.text,
  },
  {
    uid: 'count',
    label: 'Count',
    type: EFilterType.number,
    defaultValue: 7,
    step: 1,
  },
];

export const demoResults: { [k: string]: IDocument & any } = {
  a: {
    uid: 'a',
    label: 'a',
    natid: '100a',
    name: 'Foo',
    count: 3,
    multiplier: 2,
    toggle: true,
    readOnly: "Can't touch me",
    select: 'a',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  ab: {
    // Note string "4" here to check we can pass string numbers without issues
    uid: 'ab',
    label: 'ab',
    natid: '10a',
    name: 'Foobar',
    count: '99',
    def: 3,
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  b: {
    uid: 'b',
    label: 'b',
    natid: '50a',
    name: 'Bar',
    description:
      'A really really really long description that needs to fit into a little box!',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  c: {
    uid: 'c',
    label: 'c',
    name: 'C',
    count: 3,
    multiplier: 3,
    expression: 9,
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
  d: {
    uid: 'd',
    label: 'd',
    name: '',
    hiddenDemoNumber: 3,
    hiddenDemo: 'Hide me',
  },
};

export const demoHeadingsData: THeading<any>[] = [
  {
    uid: 'uid',
    label: 'UID',
    type: EFilterType.button,
    onClick: () => alert(),
    required: true,
    group: 0,
  } as IHeadingButton,
  {
    uid: 'name',
    label: 'Name',
    type: EFilterType.text,
    required: true,
    group: 0,
  },
  {
    uid: 'reference',
    label: 'Reference',
    collection: 'democollection',
    type: EFilterType.reference,
    searchFieldTargetField: 'label',
    allowAddNew: true,
  },
];

export const CustomField = ({ acceptValue, cellData, editMode, focused }) => {
  const actionFunc = useMemo(() => {
    return () => acceptValue(Math.random());
  }, [acceptValue]);

  useEffect(() => {
    if (editMode && focused) {
      actionFunc();
    }
  }, [editMode, focused, actionFunc]);
  return (
    <div className="dataTableCellData dataTableCellData-button">
      <button type="button" onClick={actionFunc}>
        {cellData}
      </button>
    </div>
  );
};

// eslint-disable-next-line no-unused-vars
export const customFilter = (value, expression, targetValue, item) => {
  if (Math.abs(value - targetValue) < 0.3) return true;
  return false;
};
