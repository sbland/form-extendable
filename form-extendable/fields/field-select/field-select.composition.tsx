import React from 'react';
import { CompositionWrapDefault } from '@form-extendable/composition-helpers';
import { EFilterType } from '@react_db_client/constants.client-types';
import { FieldSelect, TFieldSelectProps } from './field-select';
import { defaultVal, demoOptions } from './demo-data';

const onChange = () => {};
const onBlur = () => {};

const defaultProps: TFieldSelectProps = {
  type: EFilterType.select,
  label: 'Select field',
  uid: 'uid',
  unit: 'unit',
  value: defaultVal,
  options: demoOptions,
  onBlur,
  onChange,
};

export const BasicFieldSelect = () => (
  <CompositionWrapDefault height="4rem" width="8rem">
    <FieldSelect {...defaultProps} required />
  </CompositionWrapDefault>
);
