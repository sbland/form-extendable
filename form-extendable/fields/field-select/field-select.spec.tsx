import React from 'react';

import { FieldSelect, TFieldSelectProps } from './field-select';
import * as compositions from './field-select.composition';
import { defaultVal, demoOptions } from './demo-data';
import { render, screen } from '@testing-library/react';
import { EFilterType } from '@react_db_client/constants.client-types';

const onChange = jest.fn();

const defaultProps: TFieldSelectProps = {
  type: EFilterType.select,
  label: 'Select field',
  uid: 'uid',
  unit: 'unit',
  value: defaultVal,
  options: demoOptions,
  onChange,
};

describe('Field Select', () => {
  beforeEach(() => {
    onChange.mockClear();
  });
  test('Renders', () => {
    render(<FieldSelect {...defaultProps} />);
  });

  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        render(<Composition />);
      });
    });
  });
  describe('shallow renders', () => {
    test('Matches Snapshot', () => {
      render(<FieldSelect {...defaultProps} />);
      const tree = screen.debug();
      expect(tree).toMatchSnapshot();
    });
  });
});
