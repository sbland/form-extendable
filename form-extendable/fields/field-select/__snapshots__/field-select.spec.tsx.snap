// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Select shallow renders Matches Snapshot 1`] = `undefined`;
